### DISCUSSION TITLE
Modifying distribution function

### DISCUSSION THREAD

**Discussion by stepan-tsirkin** [wannier-berri/wannier-berri#221] [Ideas]
In reply to [question](https://github.com/wannier-berri/wannier-berri/issues/219#issuecomment-1354553497) by @AbhirupRK

 >I have one question though. Suppose I want to change the weight function in AHC formula altogether to say, log(f). Is there any shortcut? Even if not, can you guide me where to look for in the code?

I guess you mean an integral smth like $$L(\mu) =\int [dk] \Omega_{nk} \log[f(\mu-\epsilon_{nk})]$$  .
Currently the calculators work with Fermi-Dirac distribution and its derivatives. 
One way to do it is to evaluate with `fder=1`, which is essentially a $\delta$-function, so you get $$D(\mu) = \int [dk] \Omega_{nk} \delta(\mu-\epsilon_{nk})$$
after converging the integral you may get $$L(\mu)=\int\limits_{-\infty}^{\infty} \log[f(\mu-\mu')] D(\mu')$$
Note : it is better to use the tetrahedron method and unsmeared results for $D$

I think, this will work well, but if you want to dig deeper, you may modify the `StaticCalculator.__call__()` method, namely how to define the weights here:
https://github.com/wannier-berri/wannier-berri/blob/27a451f7e1e7890f9256ba0a142cf5de1c18003f/wannierberri/calculators/static.py#L50-L54

For each k-point there is a dictionary { (ib1,ib2):W } where W is the array (of the size of number of Fermi levels) which is assigned to the group of degenerate bands  `[ib1:ib2]` (python convention for indexing implied, of course). So, you may form this dictionary in any other way you like.  And you need to use the "tetra" mode, even if you do not actually use the tetrahedron method, because the "notetra" method below is hard-coded to use the step function and its derivatives, so it is not flexible (but fast)
https://github.com/wannier-berri/wannier-berri/blob/27a451f7e1e7890f9256ba0a142cf5de1c18003f/wannierberri/calculators/static.py#L84-L111

If you have more questions, please ask below in this discussion.

**Comment by AbhirupRK** [comment]
Thanks a lot @stepan-tsirkin for this detailed discussion. So far what I have understood from this are as follows:

- In the first part you converted the k-space Berry curvature to a function of energy. And then by performing an energy integration of that function along with a custom weight function, one may get the desired value. Well that's nice but what I would do instead is:

  First tabulate the Berry curvature and my desired weight function for a grid of k-points and then perform a numerical integration of the product of these two.

- Second part: After digging into the file `__tetrahedron.py`, I found that there is no analytic definition of the distribution function :
  $f(\epsilon)=\frac{1}{e^{(\epsilon - \mu)/kT} + 1}$
  
  Instead, only the numerical values of the occupation factor are written in terms of cases depending on the value of $\epsilon_F$. So, it will be difficult to replace such distribution function with something else.

  Please let me know if I got it right.

**Reply by stepan-tsirkin** [reply]
Of course, you can do whatever works for you. However note, that tabulation of k-dependent quantities is not implemented as effecient in WannierBerri, as the integration. And of course it requires a lot of memory in case of very dense grids. In other words, tabulation is meant mainly for visualisation purposes. 

Instead, after you convert everything into a spectral function, you can apply any distribution function you like in a second (it is just a 1d integral over energies), thus easily varying temperature or chemical potential. 

In the tetrahedron method the weights are calculated by breaking the k-space into tetrahedron, and taking the analytical integration within each tetrahedron of the T=0 distribution function, assuming that the energies change linearly within that tetrahedron. It is not easy (if possible) to take such distribution for any custom distribution function. 

However, what I meant is to substitute line 52 of static.py by any other way of generating the weights, probably not using tetrahedron method, just usung the energies on one k-point.  This way, lines 85-90 will be used directly. 

But anyway, converting into one energy function seems to me the most straightforward and efficient way, that's what we usually do to study temperature-dependent quantities.

**Reply by AbhirupRK** [reply]
Got it. Thanks a lot for your kind response sir @stepan-tsirkin .
